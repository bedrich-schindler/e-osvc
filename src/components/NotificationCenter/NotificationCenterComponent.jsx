import Box from '@material-ui/core/Box';
import PropTypes from 'prop-types';
import React from 'react';
import Alert from '@material-ui/lab/Alert';
import {
  NOTIFICATION_TYPE_AUTO_REMOVABLE,
  NOTIFICATION_TYPE_REMOVABLE,
  NOTIFICATION_TYPES,
  NOTIFICATION_VARIANTS,
} from '../../resources/notification';
import styles from './styles.scss';

const isRemovable = (notification) => [
  NOTIFICATION_TYPE_AUTO_REMOVABLE,
  NOTIFICATION_TYPE_REMOVABLE,
].includes(notification.type);

export const NotificationCenterComponent = (props) => {
  const {
    notifications,
    notificationRemove,
  } = props;

  if (notifications.length === 0) {
    return null;
  }

  return (
    <div className={styles.root}>
      {notifications.map((notification, index) => (
        <Box
          key={notification.id}
          mb={(notifications.length - 1) !== index ? 2 : 6}
        >
          <Alert
            onClose={
              isRemovable(notification)
                ? () => notificationRemove(notification.id)
                : null
            }
            variant="filled"
            severity={notification.variant}
          >
            {notification.message}
          </Alert>
        </Box>
      ))}
    </div>
  );
};

NotificationCenterComponent.defaultProps = {
  notifications: [],
};

NotificationCenterComponent.propTypes = {
  notificationRemove: PropTypes.func.isRequired,
  notifications: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired,
    type: PropTypes.oneOf(NOTIFICATION_TYPES).isRequired,
    variant: PropTypes.oneOf(NOTIFICATION_VARIANTS).isRequired,
  })),
};

export default NotificationCenterComponent;
