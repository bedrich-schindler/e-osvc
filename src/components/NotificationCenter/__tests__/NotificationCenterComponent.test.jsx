import React from 'react';
import sinon from 'sinon';
import {
  mount,
  shallow,
} from 'enzyme';
import { NotificationCenterComponent } from '../NotificationCenterComponent';

describe('rendering', () => {
  it('renders all types of info notification', () => {
    const tree = shallow(
      <NotificationCenterComponent
        isOnline
        notifications={[
          {
            id: '1',
            message: 'Info message - Auto-removable',
            type: 'autoRemovable',
            variant: 'info',
          },
          {
            id: '2',
            message: 'Info message - Removable',
            type: 'removable',
            variant: 'info',
          },
          {
            id: '3',
            message: 'Info message - Non-removable',
            type: 'nonRemovable',
            variant: 'info',
          },
        ]}
        notificationRemove={() => {}}
      />,
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders all types of success notification', () => {
    const tree = shallow(
      <NotificationCenterComponent
        isOnline
        notifications={[
          {
            id: '1',
            message: 'Success message - Auto-removable',
            type: 'autoRemovable',
            variant: 'success',
          },
          {
            id: '2',
            message: 'Success message - Removable',
            type: 'removable',
            variant: 'success',
          },
          {
            id: '3',
            message: 'Success message - Non-removable',
            type: 'nonRemovable',
            variant: 'success',
          },
        ]}
        notificationRemove={() => {}}
      />,
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders all types of warning notification', () => {
    const tree = shallow(
      <NotificationCenterComponent
        isOnline
        notifications={[
          {
            id: '1',
            message: 'Warning message - Auto-removable',
            type: 'autoRemovable',
            variant: 'warning',
          },
          {
            id: '2',
            message: 'Warning message - Removable',
            type: 'removable',
            variant: 'warning',
          },
          {
            id: '3',
            message: 'Warning message - Non-removable',
            type: 'nonRemovable',
            variant: 'warning',
          },
        ]}
        notificationRemove={() => {}}
      />,
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders all types of error notification', () => {
    const tree = shallow(
      <NotificationCenterComponent
        isOnline
        notifications={[
          {
            id: '1',
            message: 'Error message - Auto-removable',
            type: 'autoRemovable',
            variant: 'error',
          },
          {
            id: '2',
            message: 'Error message - Removable',
            type: 'removable',
            variant: 'error',
          },
          {
            id: '3',
            message: 'Error message - Non-removable',
            type: 'nonRemovable',
            variant: 'error',
          },
        ]}
        notificationRemove={() => {}}
      />,
    );

    expect(tree).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls notificationRemove when close notification button is clicked', () => {
    const spy = sinon.spy();
    const tree = mount(
      <NotificationCenterComponent
        isOnline
        notifications={[
          {
            id: '1',
            message: 'Info message - Removable',
            type: 'removable',
            variant: 'info',
          },
        ]}
        notificationRemove={spy}
      />,
    );

    tree.find('button').at(0).simulate('click');
    expect(spy.calledOnce).toEqual(true);
  });
});
