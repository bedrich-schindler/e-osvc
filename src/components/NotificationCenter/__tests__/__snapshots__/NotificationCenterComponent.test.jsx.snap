// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders all types of error notification 1`] = `
<div
  className="root"
>
  <Styled(MuiBox)
    mb={2}
  >
    <WithStyles(ForwardRef(Alert))
      onClose={[Function]}
      severity="error"
      variant="filled"
    >
      Error message - Auto-removable
    </WithStyles(ForwardRef(Alert))>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    mb={2}
  >
    <WithStyles(ForwardRef(Alert))
      onClose={[Function]}
      severity="error"
      variant="filled"
    >
      Error message - Removable
    </WithStyles(ForwardRef(Alert))>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    mb={6}
  >
    <WithStyles(ForwardRef(Alert))
      onClose={null}
      severity="error"
      variant="filled"
    >
      Error message - Non-removable
    </WithStyles(ForwardRef(Alert))>
  </Styled(MuiBox)>
</div>
`;

exports[`rendering renders all types of info notification 1`] = `
<div
  className="root"
>
  <Styled(MuiBox)
    mb={2}
  >
    <WithStyles(ForwardRef(Alert))
      onClose={[Function]}
      severity="info"
      variant="filled"
    >
      Info message - Auto-removable
    </WithStyles(ForwardRef(Alert))>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    mb={2}
  >
    <WithStyles(ForwardRef(Alert))
      onClose={[Function]}
      severity="info"
      variant="filled"
    >
      Info message - Removable
    </WithStyles(ForwardRef(Alert))>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    mb={6}
  >
    <WithStyles(ForwardRef(Alert))
      onClose={null}
      severity="info"
      variant="filled"
    >
      Info message - Non-removable
    </WithStyles(ForwardRef(Alert))>
  </Styled(MuiBox)>
</div>
`;

exports[`rendering renders all types of success notification 1`] = `
<div
  className="root"
>
  <Styled(MuiBox)
    mb={2}
  >
    <WithStyles(ForwardRef(Alert))
      onClose={[Function]}
      severity="success"
      variant="filled"
    >
      Success message - Auto-removable
    </WithStyles(ForwardRef(Alert))>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    mb={2}
  >
    <WithStyles(ForwardRef(Alert))
      onClose={[Function]}
      severity="success"
      variant="filled"
    >
      Success message - Removable
    </WithStyles(ForwardRef(Alert))>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    mb={6}
  >
    <WithStyles(ForwardRef(Alert))
      onClose={null}
      severity="success"
      variant="filled"
    >
      Success message - Non-removable
    </WithStyles(ForwardRef(Alert))>
  </Styled(MuiBox)>
</div>
`;

exports[`rendering renders all types of warning notification 1`] = `
<div
  className="root"
>
  <Styled(MuiBox)
    mb={2}
  >
    <WithStyles(ForwardRef(Alert))
      onClose={[Function]}
      severity="warning"
      variant="filled"
    >
      Warning message - Auto-removable
    </WithStyles(ForwardRef(Alert))>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    mb={2}
  >
    <WithStyles(ForwardRef(Alert))
      onClose={[Function]}
      severity="warning"
      variant="filled"
    >
      Warning message - Removable
    </WithStyles(ForwardRef(Alert))>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    mb={6}
  >
    <WithStyles(ForwardRef(Alert))
      onClose={null}
      severity="warning"
      variant="filled"
    >
      Warning message - Non-removable
    </WithStyles(ForwardRef(Alert))>
  </Styled(MuiBox)>
</div>
`;
